tools/halide_image_io.h

```diff
@@ -1306,6 +1307,240 @@ bool save_mat(ImageType &im, const std::string &filename) {
+    tag++->assign32(257, 1, height);                         // ImageLength
+    tag++->assign16(258, 1, int16_t(bytes_per_element * 8)); // BitsPerSample
+    tag++->assign16(259, 1, 1);                              // Compression -- none
+    tag++->assign16(262, 1, channels >= 3 ? 2 : 1);          // PhotometricInterpretation -- black is zero or RGB
+    tag++->assign32(273, channels, sizeof(header));          // StripOffsets
+    tag++->assign16(277, 1, int16_t(channels));              // SamplesPerPixel
+    tag++->assign32(278, 1, height);                         // RowsPerStrip
+    tag++->assign32(279, channels,                           // StripByteCounts
+                    (channels == 1) ?
+                        elements * bytes_per_element :
+                        sizeof(header) + channels * sizeof(int32_t));  // for channels > 1, this is an offset
+    tag++->assign32(282, 5, 1,
+                    offsetof(halide_tiff_header, width_resolution));     // XResolution
+    tag++->assign32(283, 5, 1,
+                    offsetof(halide_tiff_header, height_resolution));    // YResolution
+    tag++->assign16(284, 1, 2);                              // PlanarConfiguration -- planar
+    tag++->assign16(296, 1, 1);                              // ResolutionUnit -- none
+    tag++->assign16(339, 1, type_code_to_tiff_sample_type[im_type.code]);        // SampleFormat
+    tag++->assign32(32997, 1, depth);                        // Image depth
+
+    // Verify we used exactly the number we declared
+    assert(tag == &header.entries[header.entry_count]);
+
+    header.ifd0_end = 0;
+    header.width_resolution[0] = 1;
+    header.width_resolution[1] = 1;
+    header.height_resolution[0] = 1;
+    header.height_resolution[1] = 1;
+
+    if (!check(f.write_bytes(&header, sizeof(header)), "TIFF write failed")) {
+        return false;
+    }
+
+    if (channels > 1) {
+        // Fill in the values for StripOffsets
+        int32_t offset = sizeof(header) + channels * sizeof(int32_t) * 2;
+        for (int32_t i = 0; i < channels; i++) {
+            if (!check(f.write_bytes(&offset, sizeof(offset)), "TIFF write failed")) {
+                return false;
+            }
+            offset += width * height * depth * bytes_per_element;
+        }
+        // Fill in the values for StripByteCounts
+        int32_t count = width * height * depth * bytes_per_element;
+        for (int32_t i = 0; i < channels; i++) {
+            if (!check(f.write_bytes(&count, sizeof(count)), "TIFF write failed")) {
+                return false;
+            }
+        }
+    }
+
+    // Reorder the data according to the strides.
+    const int TEMP_SIZE = 4 * 1024;
+    uint8_t temp[TEMP_SIZE];
+    const int max_elements = TEMP_SIZE / bytes_per_element;
+    int counter = 0;
+    const uint8_t *base = (const uint8_t*) im.data();
+    for (int32_t dim3 = shape[3].min; dim3 < shape[3].extent + shape[3].min; ++dim3) {
```

Comment:
I know this is largely copy-pasted from the runtime, but given all the tools we have in Runtime::Buffer, it really feels like we should be able to do better than this loop. E.g. I think the version of for_each_element that uses a const int * for the coordinates would be simpler.

Would also be nice to check if the buffer is dense and just do a single write_bytes in that case.

Reply:
> the version of for_each_element

The reason this code doesn't do that is that it `for_each_element` and `for_each_value` require statically-typed `Buffer<T>`, and this function has only a `Buffer<void>`. We could of course sniff the type and dispatch to an appropriate template instantiation, but it would mean some code expansion since we'd have to instantiate for all 11 or so possible types. Is that a price worth paying?

Reply:
(PTAL)